Components/Queues/AsyncQueueConsumer.php

```diff
@@ -4,90 +4,149 @@
+        $this->driver = $driver;
     }
 
     /**
-     * Consumes messages from the given $endpoint until either the expirationCount reaches 0 or ::stop() is called.
-     *
-     * @param EndpointInterface $endpoint
-     *
-     * @throws \Exception
-     *
-     * @return bool
+     * {@inheritdoc}
      */
-    public function consume(EndpointInterface $endpoint)
+    protected function initialize(EndpointInterface $endpoint)
     {
-        $handler = new AmqpQueueHandler($endpoint, (int) $this->expirationCount, $this->format, $this->serializer);
-        $handler->setExceptionHandler($this->getExceptionHandler());
-
-        if ($this->smartesbHelper) {
-            $handler->setSmartesbHelper($this->smartesbHelper);
-        }
-        if ($this->logger) {
-            $handler->setLogger($this->logger);
-        }
-
-        try {
-            $this->manager->getQueue($this->getQueueName($endpoint))->consume($handler);
-        } finally {
-            $this->manager->disconnect();
+        if (!$this->driver->isConnected()) {
+            $this->driver->connect();
         }
-
-        return true;
     }
 
     /**
+     * Returns the consumer name.
+     *
      * {@inheritdoc}
      */
     public function getName()
     {
-        return 'async_queue_consumer';
+        return sprintf(self::CONSUMER_TAG, gethostname(), getmypid());
     }
 
+    /**
+     * Returns the queue name properly treated with queue prefix.
+     */
     protected function getQueueName(EndpointInterface $endpoint): string
     {
         $options = $endpoint->getOptions();
 
         return "{$options[QueueProtocol::OPTION_PREFIX]}{$options[QueueProtocol::OPTION_QUEUE_NAME]}";
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function cleanUp(EndpointInterface $endpoint)
+    {
+        $this->driver->disconnect();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function confirmMessage(EndpointInterface $endpoint, MessageInterface $message)
+    {
+        $this->driver->ack($message->getMessageId());
+    }
+
+    public function asyncConsume(EndpointInterface $endpoint, callable $callback)
+    {
+        $this->driver->declareChannel();
+        $queueName = $this->getQueueName($endpoint);
+
+        $this->driver->declareQueue($queueName);
+        $this->driver->consume($this->getName(), $queueName, $callback);
+    }
+
+    public function waitNoBlock()
+    {
+        if ($this->driver->isConsuming()) {
+            $this->driver->waitNoBlock();
+        }
+    }
+
+    public function wait()
+    {
+        if ($this->driver->isConsuming()) {
+            $this->driver->wait();
+        }
+    }
+
+    protected function process(EndpointInterface $queueEndpoint, QueueMessageInterface $message)
+    {
+        $endpoint = $this->smartesbHelper->getEndpointFactory()->createEndpoint($message->getDestinationURI(), EndpointFactory::MODE_CONSUME);
+        $queueEndpoint->getHandler()->handle($message->getBody(), $endpoint);
```

Comment:
Is this method able to process messages doesn't matter which which queue (like retry, throttling and so on)? In the QueueConsumer the code is like this
`
    protected function process(EndpointInterface $queueEndpoint, MessageInterface $message)
    {
        // If we used a wrapper to queue the message, that the handler doesn't understand, unwrap it
        if ($message instanceof QueueMessageInterface && !($queueEndpoint->getHandler() instanceof QueueMessageHandlerInterface)) {
            $endpoint = $this->smartesbHelper->getEndpointFactory()->createEndpoint($message->getDestinationURI(), EndpointFactory::MODE_CONSUME);
            $queueEndpoint->getHandler()->handle($message->getBody(), $endpoint);
        } else {
            parent::process($queueEndpoint, $message);
        }
    }
`

Reply:
So that's why that `if` is there! I'll put it back.